README.md
# Arctic Overlay

A fast overlay for Hypixel.

Currently is able to grab stats for Hypixel Bedwars.

Available for Windows and Mac.

**DOWNLOAD: https://github.com/TimStewartJ/ArcticOverlay/releases**

## Installing

The executables will not be signed. This means a warning will pop up regarding the trustworthiness of the application. This is because it costs a lot of money to make that warning go away. The source code is all in this repository, and the executable can simply be manually built if it's trustworthiness is questioned.

### Mac OS Notes

The Mac version of the app will ask for permission in accessibility settings. This is for the autowho functionality. The permissions are necessary to detect if the current focused app is Minecraft. If it is not Minecraft, then autowho will not be triggered.

## Updating

The Windows release will detect if theres an update and update automatically on launch.

Unfortunately, auto update is not available for Mac. Mac users will have to reinstall the overlay every time they wish to update.

## Development

The package `active-win` is used to see which window is active. One of its dependencies does not work nicely with Electron.js. Run the executable for `electron-rebuild` in `node_modules/.bin` for the appropriate platform if necessary. Also, builds are only known to work with Node.js version `16.13.0`.

## Building

Asar compressing has to be off for now so that node-key-sender can find its jar file.

CHANGELOG.md
# Changelog

## Unreleased
### Changed
 - Minor aesthetic changes
### Removed
 - Light mode

## [1.3.0] - 2022-01-13
### Changed
 - Colors are now based on set cutoffs.
 - Moved checkboxes to the left of their labels.
 - Coloring individual stats or the whole player's stats is toggleable.

## [1.2.1] - 2022-01-13
### Changed
 - Utility functions for improved runtime and readability.
### Fixed
 - File size fetching when `latest.log` gets renamed.
 - No more screen recording permission.

## [1.2.0] - 2021-11-21
### Added
 - Players who say the user's name are added to the overlay.
 - Commands in-game via `/t .`.
 - `/p list` support.
 - Players invited to the party are added to the overlay.
### Changed
 - Close button now closes settings if settings are up.
 - Line reading logic.
### Fixed
 - Icon not showing up in header.
### Removed
 - Clear table button.

## [1.1.0] - 2021-11-20
### Removed
 - Default title bar, replaced with custom one.
### Added
 - Custom title bar.
 - New app icon.
### Changed
 - Miscellaneous visuals.
 - How settings work internally.
 - Darkmode preference is now saved.

## [1.0.1] - 2021-11-19
### Added
 - Dark mode.

## [1.0.0] - 2021-11-19
### Features
 - Working Hypixel bedwars overlay.
 - Autowho.
 - Mode Switching.
 - Manual Player Lookup.


index.js
const { app, BrowserWindow, ipcMain, globalShortcut } = require('electron');
const { write, read } = require('./src/settings.js');
const pathjs = require('path');
const { readFromFile, manualLookup, refreshPlayers } = require('./src/readPlayers.js');
const { paths } = require('./data.json');
const windowStateKeeper = require('electron-window-state');
const { autoUpdater } = require('electron-updater');
try {
    require('electron-reloader')(module);
} catch {}

app.on('ready', () => {

    autoUpdater.checkForUpdatesAndNotify();

    // Load the previous state with fallback to defaults
    const mainWindowState = windowStateKeeper({
        defaultWidth: 800,
        defaultHeight: 500
    });

    // init the window and set it to load from index.html
    const win = new BrowserWindow({
        show: false,
        x: mainWindowState.x,
        y: mainWindowState.y,
        width: mainWindowState.width,
        height: mainWindowState.height,
        minWidth: 800,
        minHeight: 400,
        frame: false,
        transparent: true,
        resizable: true,
        show: false,
        // backgroundColor: '#00ffffff',
        webPreferences: {
            nodeIntegration: true,
            preload: pathjs.join(__dirname, 'src/preload.js')
        }
    });

    // this will let electron window state manage the window for us
    mainWindowState.manage(win);

    win.loadFile('index.html');

    win.removeMenu();

    win.once('ready-to-show', () => {
        win.show();
    });

    // shortcuts - f5 to refresh page and ctrl+shift+i to show dev tools
    globalShortcut.register('f5', () => {
        win.reload();
    });

    globalShortcut.register('CommandOrControl+Shift+I', () => {
        win.webContents.openDevTools();
    });

    const clientSelect = (client) => {
        write('path', `${app.getPath('home').replace(/\\/g, '\/') + paths[`${client}-${process.platform}`]  }/logs/latest.log`);
        write('client', client);
    };

    if(!read('client')) {
        clientSelect('lunar');
    }
    if(!read('autowho')) {
        write('autowho', true);
    }
    if(!read('mode')) {
        write('mode', 'overall');
    }
    if(!read('theme-darkmode')) {
        write('theme-darkmode', false);
    }
    if(!read('statColor')) {
        write('statColor', false);
    }

    win.webContents.once('dom-ready', () => {
        win.webContents.send('initSettings', {
            path: read('path'),
            client: read('client'),
            autowho: read('autowho'),
            mode: read('mode'),
            darkmode: read('theme-darkmode'),
            statColor: read('statColor')
        });
        // start reading from the file immediately
        readFromFile(read('path'), win, read('key'));
    });

    // listen for settings changes

    ipcMain.on('updateSettings', (e, data) => {
        switch(data.type) {
        case 'client':
            clientSelect(data.client);
            win.webContents.send('noticeText', 'Please restart the overlay!');
            break;
        case 'autowho':
            write('autowho', data.autowho);
            break;
        case 'modeSelect':
            write('mode', data.modeSelect.mode);
            refreshPlayers(data.modeSelect.players, win, read('key'));
            break;
        case 'darkmode':
            write('theme-darkmode', data.darkmode);
            break;
        case 'statColor':
            write('statColor', data.statColor);
            break;
        }
    });

    ipcMain.on('manualLookup', (e, data) => {
        manualLookup(data, win, read('key'));
    });

    ipcMain.on('exit', (e, data) => {
        win.close();
    });
});

app.on('window-all-closed', () => {
    app.quit();
});

src/util.js
// util.js: lots of utility data and functions
// some of this is taken from statsify code, big thanks to their team!

// objects and arrays holding data:

const xpperlevel = [500, 1000, 2000, 3500];

const colors = {
    0: { hsl: 'hsl(0, 0%, 0%)' },
    1: { hsl: 'hsl(240, 100%, 33%)' },
    2: { hsl: 'hsl(120, 100%, 33%)' },
    3: { hsl: 'hsl(180, 100%, 33%)' },
    4: { hsl: 'hsl(0, 100%, 33%)' },
    5: { hsl: 'hsl(300, 100%, 33%)' },
    6: { hsl: 'hsl(40, 100%, 50%)' },
    7: { hsl: 'hsl(0, 0%, 67%)' },
    8: { hsl: 'hsl(0, 0%, 33%)' },
    9: { hsl: 'hsl(240, 100%, 67%)' },
    a: { hsl: 'hsl(120, 100%, 67%)' },
    b: { hsl: 'hsl(180, 100%, 67%)' },
    c: { hsl: 'hsl(0, 100%, 67%)' },
    d: { hsl: 'hsl(300, 100%, 67%)' },
    e: { hsl: 'hsl(60, 100%, 67%)' },
    f: { hsl: 'hsl(0, 0%, 100%)' },
    g: { hsl: 'hsl(58, 96%, 44%)' },
    s: { hsl: 'hsl(205, 90%, 56%)' },
};

const prestigeColors = [
    { req: 0, fn: (n) => `§7[${n}✫]` },
    { req: 100, fn: (n) => `§f[${n}✫]` },
    { req: 200, fn: (n) => `§6[${n}✫]` },
    { req: 300, fn: (n) => `§b[${n}✫]` },
    { req: 400, fn: (n) => `§2[${n}✫]` },
    { req: 500, fn: (n) => `§3[${n}✫]` },
    { req: 600, fn: (n) => `§4[${n}✫]` },
    { req: 700, fn: (n) => `§d[${n}✫]` },
    { req: 800, fn: (n) => `§9[${n}✫]` },
    { req: 900, fn: (n) => `§5[${n}✫]` },
    {
        req: 1000,
        fn: (n) => {
            const nums = n.toString().split('');
            return `§c[§6${nums[0]}§e${nums[1]}§a${nums[2]}§b${nums[3]}§d✫§5]`;
        },
    },
    { req: 1100, fn: (n) => `§7[§f${n}§7✪]` },
    { req: 1200, fn: (n) => `§7[§e${n}§6✪§7]` },
    { req: 1300, fn: (n) => `§7[§b${n}§3✪§7]` },
    { req: 1400, fn: (n) => `§7[§a${n}§2✪§7]` },
    { req: 1500, fn: (n) => `§7[§3${n}§9✪§7]` },
    { req: 1600, fn: (n) => `§7[§c${n}§4✪§7]` },
    { req: 1700, fn: (n) => `§7[§d${n}§5✪§7]` },
    { req: 1800, fn: (n) => `§7[§9${n}§1✪§7]` },
    { req: 1900, fn: (n) => `§7[§5${n}§8✪§7]` },
    {
        req: 2000,
        fn: (n) => {
            const nums = n.toString().split('');
            return `§8[§7${nums[0]}§f${nums[1]}${nums[2]}§7${nums[3]}§8✪]`;
        },
    },
    {
        req: 2100,
        fn: (n) => {
            const nums = n.toString().split('');
            return `§f[${nums[0]}§e${nums[1]}${nums[2]}§6${nums[3]}⚝]`;
        },
    },
    {
        req: 2200,
        fn: (n) => {
            const nums = n.toString().split('');
            return `§6[${nums[0]}§f${nums[1]}${nums[2]}§b${nums[3]}§3⚝]`;
        },
    },
    {
        req: 2300,
        fn: (n) => {
            const nums = n.toString().split('');
            return `§5[${nums[0]}§d${nums[1]}${nums[2]}§6${nums[3]}§e⚝]`;
        },
    },
    {
        req: 2400,
        fn: (n) => {
            const nums = n.toString().split('');
            return `§b[${nums[0]}§f${nums[1]}${nums[2]}§7${nums[3]}⚝§8]`;
        },
    },
    {
        req: 2500,
        fn: (n) => {
            const nums = n.toString().split('');
            return `§f[${nums[0]}§a${nums[1]}${nums[2]}§2${nums[3]}⚝]`;
        },
    },
    {
        req: 2600,
        fn: (n) => {
            const nums = n.toString().split('');
            return `§4[${nums[0]}§c${nums[1]}${nums[2]}§d${nums[3]}5]`;
        },
    },
    {
        req: 2700,
        fn: (n) => {
            const nums = n.toString().split('');
            return `§e[${nums[0]}§f${nums[1]}${nums[2]}§8${nums[3]}⚝]`;
        },
    },
    {
        req: 2800,
        fn: (n) => {
            const nums = n.toString().split('');
            return `§a[${nums[0]}§2${nums[1]}${nums[2]}§6${nums[3]}⚝§e]`;
        },
    },
    {
        req: 2900,
        fn: (n) => {
            const nums = n.toString().split('');
            return `§b[${nums[0]}§3${nums[1]}${nums[2]}§9${nums[3]}⚝§1]`;
        },
    },
    {
        req: 3000,
        fn: (n) => {
            const nums = n.toString().split('');
            return `§e[${nums[0]}§6${nums[1]}${nums[2]}§c${nums[3]}⚝§4]`;
        },
    },
];

// general functions

exports.ratio = (top = 0, bottom = 1) => {
    return (top / bottom);
};

const tieredColors = [[170, 170, 170], [255,255,255], [255, 170, 0], [0, 170, 170], [170, 0, 0], [170, 0, 170]];

exports.color = (top, bottom, val) => {
    const norm = 25/top;
    val *= norm;
    return tieredColors[Math.min(tieredColors.length - 1, Math.floor(Math.sqrt(val)))];
};

exports.getBwLevel = (exp = 0) => {
    const prestiges = Math.floor(exp / 487000);
    let level = prestiges * 100;
    let remainingExp = exp - prestiges * 487000;

    for (let i = 0; i < 4; ++i) {
        const expForNextLevel = xpperlevel[i];
        if (remainingExp < expForNextLevel) break;
        level++;
        remainingExp -= expForNextLevel;
    }

    return parseFloat((level + remainingExp / 5000).toFixed(2));
};

// function to convert mc colors into something usable via html
exports.mcColor = (text) => {
    let finalText = '';
    splitText = text.split('§').slice(1);

    splitText.forEach(part => {
        const color = colors[part[0]].hsl || '';

        finalText += `<span style="color: ${color};">${part.substring(1)}</span>`;
    });
    return finalText;
};

// function that gives mc color based on star level.
exports.getBwFormattedLevel = (star) => {
    return prestigeColors[Math.floor(star/100)].fn(star);
};

// RANK STUFF

const ranks = {
    SUPERSTAR: 'MVP++',
    VIP_PLUS: 'VIP+',
    MVP_PLUS: 'MVP+',
    MVP: 'MVP',
    VIP: 'VIP',
    NONE: 'NON',
};

const rankColors = {
    'MVP+': { mc: '§c', hex: '#FF5555' },
    'MVP++': { mc: '§c', hex: '#FFAA00' },
    'VIP+': { mc: '§6', hex: '#FFAA00' },
    'PIG+++': { mc: '§b', hex: '#FF55FF' },
};

const rankColorsPlus = {
    RED: { mc: '§c', hex: '#FF5555' },
    GOLD: { mc: '§6', hex: '#FFAA00' },
    GREEN: { mc: '§a', hex: '#55FF55' },
    YELLOW: { mc: '§e', hex: '#FFFF55' },
    LIGHT_PURPLE: { mc: '§d', hex: '#FF55FF' },
    WHITE: { mc: '§f', hex: '#F2F2F2' },
    BLUE: { mc: '§9', hex: '#5555FF' },
    DARK_GREEN: { mc: '§2', hex: '#00AA00' },
    DARK_RED: { mc: '§4', hex: '#AA0000' },
    DARK_AQUA: { mc: '§3', hex: '#00AAAA' },
    DARK_PURPLE: { mc: '§5', hex: '#AA00AA' },
    DARK_GRAY: { mc: '§8', hex: '#555555' },
    BLACK: { mc: '§0', hex: '#000000' },
    DARK_BLUE: { mc: '§1', hex: '#0000AA' },
};

exports.getFormattedRank = (json) => {
    let rank = json.rank;
    if (!rank || rank == 'NORMAL')
    {
        if(json.monthlyPackageRank == 'NONE') json.monthlyPackageRank = null;
        rank = ranks[json.monthlyPackageRank || json.newPackageRank || json.packageRank || 'NONE'];
    }
    if (json.prefix) rank = json.prefix.replace(/§.|\[|]/g, '');

    const plus = json.rankPlusColor;
    let color;
    if (plus == undefined || rank == 'PIG+++') {
        color = rankColors[rank] || { mc: '§7', hex: '#BAB6B6' };
    } else {
        color = rankColorsPlus[plus] || { mc: '§7', hex: '#BAB6B6' };
    }
    color = color.mc;

    rank = {
        'MVP+': `§b[MVP${color}+§b]`,
        'MVP++': `§6[MVP${color}++§6]`,
        MVP: '§b[MVP]',
        'VIP+': `§a[VIP${color}+§a]`,
        VIP: '§a[VIP]',
        YOUTUBER: '§c[§fYOUTUBE§c]',
        'PIG+++': `§d[PIG${color}+++§d]`,
        HELPER: '§9[HELPER]',
        MODERATOR: '§2[MOD]',
        ADMIN: '§c[ADMIN]',
        OWNER: '§c[OWNER]',
        SLOTH: '§c[SLOTH]',
        ANGUS: '§c[ANGUS]',
        APPLE: '§6[APPLE]',
        MOJANG: '§6[MOJANG]',
        'BUILD TEAM': '§3[BUILD TEAM]',
        EVENTS: '§6[EVENTS]',
    }[rank];

    if (!rank) return '§7';
    return `${rank} `;
};


src/preload.js
const { contextBridge, ipcRenderer, ipcMain } = require('electron');

contextBridge.exposeInMainWorld('settings', {
    initSettings: (event, func) => ipcRenderer.once('initSettings', (event, data) => func(data)),
    noticeText: (event, func) => ipcRenderer.on('noticeText', (event, data) => func(data)),
    updateSettings: (data) => ipcRenderer.send('updateSettings', data),
});

contextBridge.exposeInMainWorld('players', {
    display: (event, func) => ipcRenderer.on('showPlayers', (event, data) => func(data)),
    add: (event, func) => ipcRenderer.on('addPlayer', (event, data) => func(data)),
    update: (event, func) => ipcRenderer.on('updatePlayer', (event, data) => func(data)),
    delete: (event, func) => ipcRenderer.on('deletePlayer', (event, data) => func(data))
});

contextBridge.exposeInMainWorld('misc', {
    manualLookup: (data) => ipcRenderer.send('manualLookup', data),
    exit: () => ipcRenderer.send('exit'),
});

src/renderer.js
const fields = ['PLAYER', 'TAG', 'WS', 'FKDR', 'WLR', 'BBLR', 'FINAL'];
const ratios = ['FKDR', 'WLR', 'BBLR'];

let mode;

let playerData = {};

let statColor = false;

const sortTable = (tableName) => {
    console.time('sort');
    const table = document.getElementById(tableName);
    let rows, i, x, y, shouldSwitch;
    let switching = true;
    while(switching) {
        switching = false;
        rows = table.rows;
        for (i = 1; i < (rows.length - 1); i++) {
            // Start by saying there should be no switching:
            shouldSwitch = false;

            /* Get the two elements you want to compare,
            one from current row and one from the next: */
            x = rows[i].getAttribute('id');
            y = rows[i + 1].getAttribute('id');

            // Check if the two rows should switch place:
            if (playerData[x].sortPriority < playerData[y].sortPriority) {
                // If so, mark as a switch and break the loop:
                shouldSwitch = true;
                break;
            }
        }
        if (shouldSwitch) {
        /* If a switch has been marked, make the switch
        and mark that a switch has been done: */
            rows[i].parentNode.insertBefore(rows[i + 1], rows[i]);
            switching = true;
        }
    }
    console.timeEnd('sort');
};

// table updating functions from main

// create a new row for the player and populate it with the default ...
const createPlayerRow = (name) => {
    const node = document.createElement('tr');
    node.setAttribute('id', name.toLowerCase());
    node.classList.add('player');
    fields.forEach(field => {
        if(field === 'PLAYER') node.innerHTML += `<td class="player-name">${name}</td>`;
        else node.innerHTML += '<td class=\"centered\">...</td>';
    });

    // add blank player to playerData
    playerData[name.toLowerCase()] = {sortPriority: -1};

    return node;
};

// reset the table and show the players in the provided input
const showPlayers = (players) => {
    const table = document.getElementById('main-table');

    table.innerHTML = ''; // reset table
    playerData = {}; // reset player data structure

    // create the header row and add the field titles to it
    const headrow = document.createElement('tr');
    fields.forEach(field => {
        let classText = '';
        if(field === 'PLAYER') classText = ' class="playerName"';
        headrow.innerHTML += `<th${classText}>${field}</th>`;
    });
    table.appendChild(headrow);

    players.forEach(player => { // for each player, create a row and populate each field with ...
        table.appendChild(createPlayerRow(player));
    });
};

window.players.display('showPlayers', (players) => {
    showPlayers(players);
});

window.players.update('updatePlayer', (data) => {
    // if we are receiving sniper data
    if(data.sniperData) {
        $(`#${data.user.toLowerCase()} td:nth-child(${2})`).text('SNIPER');
        return;
    }

    const playerRow = document.getElementById(data.user.toLowerCase()); // find the player row
    playerRow.innerHTML = ''; // reset the player row's data

    playerData[data.user.toLowerCase()] = data; // set the player's data in the dictionary

    //if the player is not a nick, color them according to their threat level
    if(!data.nick) {
        // if we are on statsify color mode
        if(statColor)
        {
            const rgb = data.stats.bedwars[mode].colors.overall;
            $(`#${data.user.toLowerCase()}`).css('color', `rgb(${rgb[0]},${rgb[1]},${rgb[2]})`);
        }
        
        playerData[data.user.toLowerCase()].sortPriority = data.stats.bedwars[mode].fkdr;
    }
    else {
        playerRow.classList.add('nick');
        playerData[data.user.toLowerCase()].sortPriority = 1000;
    }

    // populate each field with the necessary data
    fields.forEach(field => {
        const fieldNode = document.createElement('td');
        let fieldContent;
        switch(field.toLowerCase()) {
        case 'player':
            fieldContent = `${data.nick ? `${data.user}` : `${data.coloredstar} ${data.displayName}`}`;
            break;
        case 'tag':
            fieldContent = data.nick ? 'NICK' : '';
            break;
        default:
            fieldContent = data.nick ? 'NICK' : ratios.includes(field) ? data.stats.bedwars[mode][field.toLowerCase()].toFixed(2) : data.stats.bedwars[mode][field.toLowerCase()];
            if(!data.nick && !statColor)
            {
                const fieldColor = data.stats.bedwars[mode].colors[field.toLowerCase()];
                fieldNode.style.color = `rgb(${fieldColor[0]},${fieldColor[1]},${fieldColor[2]})`;
            }   
            break;
        }
        fieldNode.innerHTML = fieldContent;
        field !== 'PLAYER' ? fieldNode.classList.add('centered') : fieldNode.classList.add('player-name');
        playerRow.appendChild(fieldNode);
    });
    sortTable('main-table'); // re sort the table
});

window.players.add('addPlayer', (player) => {
    if($(`#${player.toLowerCase()}`).length) return;
    const table = document.getElementById('main-table');
    table.appendChild(createPlayerRow(player));
});

window.players.delete('deletePlayer', (player) => {
    // find the row corresponding to the player, and if it exists, delete it
    const row = document.getElementById(player.toLowerCase());
    if(row) row.parentNode.removeChild(row);

    // remove the player from playerData
    delete playerData[player.toLowerCase()];
});

// settings functions

window.settings.initSettings('initSettings', (settings) => {
    $(`#${settings.client}`).prop('selected', true);
    $('#autowho').prop('checked', settings.autowho);
    // $('#darkmode').prop('checked', settings.darkmode);
    $('#statColor').prop('checked', settings.statColor);
    $(`#${settings.mode}`).prop('selected', true);
    mode = settings.mode;
    statColor = settings.statColor;
    themeChange(true);
});

window.settings.noticeText('noticeText', (text) => {
    $('#notice-text').text(text);
});

let settingsClicked = false;
let transitioning = false;

const settingsTransition = () => {
    const transitionLength = 500;
    if(!transitioning) {
        transitioning = true;
        if(settingsClicked) {
            $('#settings-button').removeClass('clicked');
            $('#settings-button').css('transform','rotate(0deg)');
            $('#settings').animate({top: '-15em'}, transitionLength);
        }
        else {
            $('#settings-button').addClass('clicked');
            $('#settings-button').css('transform','rotate(180deg)');
            $('#settings').animate({top: '5em'}, transitionLength);
        }
    }
    setTimeout(() => {
        transitioning = false;
        settingsClicked = !settingsClicked;
    }, transitionLength);
};

$('#settings-button').click(() => {
    settingsTransition();
});

$('#client-select').change(async () => {
    await window.settings.updateSettings({ type: 'client', client: $('#client-select :selected').val() });
});

$('#autowho').change(async (data) => {
    await window.settings.updateSettings({ type: 'autowho', autowho: data.target.checked });
});

$('#mode-select').change(async () => {
    mode = $('#mode-select :selected').val();
    const players = [];
    $('.player').each(function () {
        players.push($(this).attr('id'));
    });
    await window.settings.updateSettings({ type: 'modeSelect', modeSelect: {
        mode: mode,
        players: players,
    }});
});

const themeChange = (darkmode) => {
    if(darkmode) {
        document.body.classList.add('dark-mode');
        document.getElementById('misc-footer').classList.add('dark-mode');
        document.body.classList.remove('light-mode');
        document.getElementById('misc-footer').classList.remove('light-mode');
    }
    else {
        document.body.classList.remove('dark-mode');
        document.getElementById('misc-footer').classList.remove('dark-mode');
        document.body.classList.add('light-mode');
        document.getElementById('misc-footer').classList.add('light-mode');
    }
};

$('#darkmode').change(async (data) => {
    themeChange(data.target.checked);
    await window.settings.updateSettings({ type: 'darkmode', darkmode: data.target.checked });
});

$('#statColor').change(async (data) => {
    statColor = data.target.checked;
    await window.settings.updateSettings({ type: 'statColor', statColor: data.target.checked });
});

// misc functions

$('#lookup-form').submit((e) => {
    e.preventDefault();
    const lookupName = $('#manual-lookup').val();
    $('#manual-lookup').val('');
    window.misc.manualLookup(lookupName);
});

$('#exit-button').click(async () => {
    if(settingsClicked) settingsTransition();
    else window.misc.exit();
});

src/readPlayers.js
const { read, write } = require('./settings');
const { fetchPlayer, validKey, getDisplayName, fetchSniperStatus } = require('./fetchPlayers.js');
const activeWindow = require('active-win');
const ks = require('node-key-sender');

// this will call upon the api to get the player data and then update it on the front end
const fetchAndUpdatePlayer = async (player, win, key) => {
    if(key) {
        const playerData = await fetchPlayer(player.toLowerCase(), key);
        if(!playerData.error) win.webContents.send('updatePlayer', playerData);
        const sniperData = await fetchSniperStatus(player, key);
        if(!sniperData.error && sniperData.sniper) win.webContents.send('updatePlayer', sniperData);
    }
};

exports.manualLookup = async (player, win, key) => {
    win.webContents.send('addPlayer', player);
    fetchAndUpdatePlayer(player, win, key);
};

exports.refreshPlayers = async (players, win, key) => {
    win.webContents.send('showPlayers', players);
    players.forEach( async (player) => {
        fetchAndUpdatePlayer(player, win, key);
    });
};

exports.readFromFile = async (path, win, key) => {
    // this is where the latest.log is read
    // we used to use fs's watchfile to watch with a 20ms interval and grab the last line using read-last-lines
    // because multiple lines could be written in the same moment, we now manually read off the newest bytes with fs
    // this is the fastest method so far, but further experimentation is required
    // the package `always-tail` works but is not very fast
    // maybe use the package 'tail'?

    const buffSize = 2056;

    const fs = require('fs');

    
    ks.setOption('startDelayMillisec', 25);

    const keyfetch = await validKey(key);

    let user;
    
    if(!read('key') || !keyfetch.success) {
        win.webContents.send('noticeText', 'Invalid key! Run /api new.');
    }
    else {
        user = await getDisplayName(key, keyfetch.record.owner);
    }

    let autowho = false;

    let logData;
    try {
        logData = fs.statSync(path);
    }
    catch(err) {
        win.webContents.send('noticeText', 'INVALID CLIENT - make sure your client is set properly and restart the overlay!');
        return err;
    }

    let filesize = logData.size;

    let file;

    fs.open(path, 'r', (err, fd) => {
        file = fd;
    });

    fs.watchFile(path, {interval: 20}, () => {
        fs.read(file, Buffer.alloc(buffSize), 0, buffSize, filesize, (err, bytecount, buff) => {
            filesize += bytecount;
            const lines = buff.toString().split(/\r?\n/).slice(0, -1);
            lines.forEach(line => process(line));
        });
    });

    fs.watch(path.split('/').slice(0, -1).join('/'), (event, filename) => { // watching for if latest.log get rotated olr anything
        if (event === 'rename' && filename === 'latest.log') {
            console.log(event);
            console.log(filename);
            fs.open(path, 'r', (err, fd) => {
                file = fd;
                logData = fs.statSync(path);
                filesize = logData.size;
            });
        }
    });

    const process = async (line) => {
        const temp = line.split(' [CHAT] ');
        if(temp.length === 1) return;
        line = temp[1];
        const colons = line.split(':').length - 1;
        console.log(`${colons}: ${line}`);
        if(colons === 0) {
            if (line.includes(' has joined ')) { // case for someone joining the lobby
                if(read('autowho') && (!autowho || line.includes(`${user} has joined`))) {
                    autowho = true;
                    activeWindow({screenRecordingPermission: false}).then((result)=>{
                        console.log(result);
                        if((result.owner.name).toLowerCase().includes('java') || (result.owner.name).toLowerCase().includes('minecraft')) {
                            ks.startBatch()
                                .batchTypeKey('control') // We send these keys before because they can often interfere with `/who` if they were already pressed down. Might (try) to make this configurable (somehow) if enough people use different layouts for it to matter.
                                .batchTypeKey('w')
                                .batchTypeKey('a')
                                .batchTypeKey('s')
                                .batchTypeKey('d')
                                .batchTypeKey('space')
                                .batchTypeKey('slash', 50)
                                .batchTypeKeys(['w','h','o','enter'])
                                .sendBatch();
                        }
                    });
                }
                const player = line.split(' has joined')[0];
                win.webContents.send('addPlayer', player);
                fetchAndUpdatePlayer(player, win, key);
            }
            else if (line.includes(' has quit!')) { // case for someone quiting the lobby
                const player = line.split(' has quit!')[0];
                win.webContents.send('deletePlayer', player);
            }
            else if (line.includes('FINAL KILL!')) {
                const player = line.split(' ')[0];
                win.webContents.send('deletePlayer', player);
            }
            else if (line.includes('disconnected')) {
                const player = line.split(' ')[0];
                win.webContents.send('deletePlayer', player);
            }
            else if (line.includes('reconnected.')) {
                const player = line.split(' ')[0];
                win.webContents.send('addPlayer', player);
                fetchAndUpdatePlayer(player, win, key);
            }
            else if (line.includes('Sending you to mini')) {
                win.webContents.send('showPlayers', []); // reset the front end when we join a new lobby
                autowho = false;
            }
            else if (line.includes('Your new API key is ')) {
                key = line.split('Your new API key is ')[1];
                write('key', key);
                if(await validKey(key)) win.webContents.send('noticeText', '');
            }
            // stuff for other party commands
            else if(line.includes(' to the party! They have 60 seconds to accept.')) {
                const player = line.match(/\S*(?=( to the party! They have 60 seconds to accept.))/)[0];
                win.webContents.send('addPlayer', player);
                fetchAndUpdatePlayer(player, win, key);
            }
            else if (line.includes('Can\'t find a player by the name of \'.')) {
                const command = line.split('Can\'t find a player by the name of \'.')[1].replace('\'','').replace('.','');
                switch(command) {
                case 'clear':
                    win.webContents.send('showPlayers', []);
                    break;
                default:
                    win.webContents.send('addPlayer', command);
                    fetchAndUpdatePlayer(command, win, key);
                    break;
                }
            }
        }
        else if(colons === 1) {
            if(line.includes('ONLINE: ')) { // case for /who
                const players = line.split('ONLINE: ')[1].split(', ');
                win.webContents.send('showPlayers', players);
                players.forEach( async (player) => {
                    fetchAndUpdatePlayer(player, win, key);
                });
            }
            // stuff for p list
            else if(line.includes('Party Leader: ')) {
                let player = line.split(':')[1].split(' ');
                player = player[1].includes('[') ? player[2] : player[1];
                win.webContents.send('showPlayers', [player]);
                fetchAndUpdatePlayer(player, win, key);
            }
            else if(line.includes('Party Members: ') || line.includes('Party Moderators: ')) {
                const players = line.split(':')[1].split('?').slice(0, -1);
                players.forEach(player => {
                    player = player.split(' ');
                    console.log(player);
                    player = player[1].includes('[') ? player[2] : player[1];
                    win.webContents.send('addPlayer', player);
                    fetchAndUpdatePlayer(player, win, key);
                });
            }
            // if someone says the user's name
            else if(line.includes(user)) {
                const player = line.split(':')[0].split(' ').slice(-1)[0].match(/\w+/)[0]; // this extracts the user who typed in chat
                win.webContents.send('addPlayer', player);
                fetchAndUpdatePlayer(player, win, key);
            }
        }
    };
};